sql/src/main/java/io/crate/analyze/where/WhereClauseAnalyzer.java

```diff
@@ -70,100 +65,34 @@ public static WhereClause bindAndAnalyze(WhereClause where,
-            if (table.clusteredBy() != null) {
-                clusterdIdx = table.primaryKey().indexOf(table.clusteredBy());
-                if (clusterdIdx >= 0) {
-                    clusteredBy = new HashSet<>(pkValues.size());
-                    for (List<Symbol> row : pkValues) {
-                        clusteredBy.add(row.get(clusterdIdx));
+                DocTableInfo table = ((DocTableRelation) tableRelation).tableInfo();
+                if (table.isPartitioned()) {
+                    if (table.partitions().isEmpty()) {
+                        return WhereClause.NO_MATCH;
                     }
-                }
-            }
-            List<Integer> partitionsIdx = null;
-            if (table.isPartitioned()) {
-                partitionsIdx = WhereClauseOptimizer.getPartitionIndices(table.primaryKey(), table.partitionedBy());
-            }
-            docKeys = new DocKeys(pkValues, versionInQuery, clusterdIdx, partitionsIdx);
-        } else {
-            clusteredBy = getClusteredByLiterals(query, eqExtractor, transactionContext);
-        }
+                    PartitionResult partitionResult = resolvePartitions(query, table, functions, transactionContext);
+                    if (!where.partitions().isEmpty()
+                        && !partitionResult.partitions.isEmpty()
+                        && !partitionResult.partitions.equals(where.partitions())) {
 
-        List<String> partitions = null;
-        if (table.isPartitioned() && docKeys == null) {
-            if (table.partitions().isEmpty()) {
-                return WhereClause.NO_MATCH;
-            }
-            PartitionResult partitionResult = resolvePartitions(query, table, functions, transactionContext);
-            partitions = partitionResult.partitions;
-            query = partitionResult.query;
-        }
-        return new WhereClause(query, docKeys, partitions, clusteredBy);
-    }
-
-    private Set<Symbol> getClusteredByLiterals(Symbol query, EqualityExtractor ee, TransactionContext transactionContext) {
-        if (table.clusteredBy() != null) {
-            List<List<Symbol>> clusteredValues = ee.extractParentMatches(
-                ImmutableList.of(table.clusteredBy()), query, transactionContext);
-            if (clusteredValues != null) {
-                Set<Symbol> clusteredBy = new HashSet<>(clusteredValues.size());
-                for (List<Symbol> row : clusteredValues) {
-                    clusteredBy.add(row.get(0));
+                        throw new IllegalArgumentException("Given partition ident does not match partition evaluated from where clause");
+                    }
+                    return new WhereClause(
+                        partitionResult.query,
+                        partitionResult.partitions,
+                        clusteredBy
+                    );
                 }
-                return clusteredBy;
+                return new WhereClause(query, where.partitions(), clusteredBy);
```

Comment:
Suggestion - personal preference: Is it possible to avoid the multiple returns by assigned to a variable and always return that? I feel this logic is somehow mixed in the method.

Reply:
Currently `where` is effectively final. I'd like to keep it that way - each branch has a return.